render/tests/test-attributes.js

```diff
@@ -517,13 +519,13 @@ o.spec("attributes", function() {
 	o.spec("select.value", function() {
 		function makeSelect(value) {
 			var attrs = (arguments.length === 0) ? {} : {value: value}
-			return {tag: "select", attrs: attrs, children: [
-				{tag:"option", attrs: {value: "1"}},
-				{tag:"option", attrs: {value: "2"}},
-				{tag:"option", attrs: {value: "a"}},
-				{tag:"option", attrs: {value: "0"}},
-				{tag:"option", attrs: {value: ""}}
-			]}
+			return m("select", attrs,
```

Comment:
Constructing the select element itself via hyperscript rather than manually triggers these errors https://travis-ci.org/github/MithrilJS/mithril.js/builds/767047585#L293-L311:
```
attributes > select.value > updates with the same value do not re-set the attribute if the select has focus:
1
  should equal
0
    at /home/travis/build/MithrilJS/mithril.js/render/tests/test-attributes.js:639:35
    
attributes > select.value > updates with the same value do not re-set the attribute if the select has focus:
1
  should equal
0
    at /home/travis/build/MithrilJS/mithril.js/render/tests/test-attributes.js:644:35
    
attributes > select.value > updates with the same value do not re-set the attribute if the select has focus:
2
  should equal
1
    at /home/travis/build/MithrilJS/mithril.js/render/tests/test-attributes.js:649:35
 ```

Reply:
Trying to isolate the issue from the DOM mocks, I can't replicate it with real DOM. Using a spy trap on the HTMLInputElement value setter, both the manually constructed select and the hyperscript equivalent fail the first test but none of the others:

[Flems](https://flems.io/#0=N4IgtglgJlA2CmIBcBWAjAGhAMwggzsgNqgB2AhmIkiAHQAWALmLCFgMYD2pj8PyIEAF8MZStToArQh269+NbAFdS7RhG4ACMOQDW8AMrwEagBQA3crCXwAlJuAAdUps2WATpvKNG7-JoBeTVNydwBzJSoefFoEUjDGekCAoIAGewB+ByFNJAdLa3g8gpshZ1d3eEYldxcwU0cQfGN4NUaMLx8-DHLXbQaQTgAHdW52-KsbPMa0RqFbHpc++sbh0dJx4BKizUaAJjmF3tcVwZGNDbYJwumQckPFvv7V87GrrcmdxtSH4+ez9abba3Q69WzOMobUjKVTrbR6QwtMzbexOJYeTq+fxBELhSJ8RgxOIJJIpNKZbK5a5TNyfSEVKo1FzARjkMK3ZomRjjbxYvK87qadj0PBQSqkPJEP6nNYXIGfW6zYRHJYnAayt4dD43XYgA7Kx7LdWvS5a4G6+4G6XGwHvc3fX6q-4a03Ur4gUFLAC6kMhzk4tHwQ1aA05rUYtG24xhaguplsaNcnAGSiGUG88H8AHcIIlNIl4Jp8OJaYVNFBOJpSJxGJpKgBaZq1guY9wQABGSl4mgg2Hz9ELYbUmno5H82E47CU+GjKlj3HjDj+XFI+Fr2yMPngniCAaDAE9TDH1hZPgm-q4AErwbBcwNVUyJCD4DqNKNXFF-eZ-P4AeXbkjhrQUA3hApDwAACu4wzbowh4ABIACoALIADIAJKkEMXYAKIIFEEZDNBjA1vuwaviA75ahemhhFUeTHnGqI0Qy1S1KWNg0SINFNsUnybrw7iGk8fD4u45DtggeS+DYwmaN+UJPBi5CBPC+hGFy8Z-Bi7aqTo6lIowAxKuCToYuwekIhp4amGgplKaE5aWQZmn7I0pnSrQ4oge4pgVlOBG0O2nBQPuHTkPZfTkMBnBgLQE5TvgWmKX0yYblUgm0OwViwAAwpwKiMLYtDwAAjkoVhJekfzJtFFZxSiJXlZVxnuc4nnedufmTviPBBSFYWaO2HlOml-EZduWU5flhXFWVFWwFVkVJqY7YxQ1Z5NQtSUzG1KUnF5fA+d1AUEv1oUdOwI1PGNhQCZN2WwHlBU8HNzWLaY1WjaY7DrZGm3zS1u0gNdyyHaQx3+b1EbBRd5agyt6Vbu4U1PTNr1bS1dk1X5f2NYDH1uSDvQKQp-peSo8ZsCAXBgEMeDbgI7YScY1NDushA0HsADMSAAJx88IoggBQVACFl+AyDTcgEgIwhelgsBgbonMkCL4gCJAiRtrAtDSNTNSsDQTCMEM+BIAA9BbKhDLoYRZbFFta-QOsAAJ7LQqS0AALE7uYu3getS3BwYCPg7BtiMQtiGLNCcEGrQG+4RsgCbZuW9bWF2w7YAW-HwbsK73u0GgJd5wnv361gIcSOHkeMPLQhAA)

Reply:
The spy [was buggy](https://flems.io/#0=N4IgtglgJlA2CmIBcB2ADAOgGwEYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQIAL54a9RswBW5XgKEimhAK7UebCAIAEYWgGt4AZXgJtACgButWOvgBKPcAA61PXrsAnPbTY2H1I9AF49K1ofAHN1BkFSDARqaLYWMNDwtBcAfldxPSRXOwd4IpLHcQ8vH3g2dR9PMCs3EFILeG1WvH9A4LxqrwMWkD4ABx0BbuL7RyLWnFbxJwHPIebW8cnqaeAKsr1WgCYllcGvDdGJ3R3uGdL5kFpT1aHhzeupu73Zg9a0F7nd5Xba7faPU6DJweKo7agaLTbAzGMwdaz7FzuNa+XpBELhSIxOLCNiJZKpdKZLK5fKFe5zby-WE1OoNTzANi0aKPdqWNjTAJ4oqC-p6HgsEhQWrUIoUIGXLY3MG-R6LCRnNYXEaKr49H4PQ4gE7q17rbWfW568GG54m+Xm0Hfa3-QGa4E6y30v4gSFrAC6sNhHj4GFIY06I15nTYGH20wR2huVicWK8fBG6jGUAC8BCAHcIGk9Gl4HpSDJGaU9FA+HpqHw2HpagBadqNku4nwQABG6iEegghGLLFLUe0ehYtBChD4PHUpHjmkTAmTriB-GopEbwGidR6bYK4QA8t25NGMLu2Ee89QAAo+cbwHxsACeABFczwuxM+D4rAAJAAVABZAAZcw+QAUQQeIYzGB82AbF9wx6Vo4xAaE3Q3LdK0ccxAifMI9BDMMXysBNtlsX4UyBLw2wwHh7FgKw0ggUgegxIFliBE8z20DAoHgYhqHge9H2fMigLAiDo2g+BYIweCGyQlDDXQvVaL0S8ikvU03jbcpfnwoQfD0oZhGJHxaG7BAiiCRwzO4uE3hxWgiMMEwZOsTCXKiPRu3clEvLYEY1R8oYcR4QLPLREKcHCrwcSgaLUT5EZjUw+UMGlQS-xrOcFO7PgoBfHpaASvQAHpKv8AS+DADAZznUhkw8IF032YynwYpiAGE+E0NgnAweAAEd1HsFrsnayI6oajERvGybQtaTK3QanKnysfLiUEDAipKnpuzWt4OqMuoTJ62BYH6wbhrGibYCmir027ObY2oxbHpahZVra9bsuEXLttnXaYwO0qxROoYztKLqfCum6BsEe6lqeqxprddMeHehaHuW36MP+t4NqBradsK4rIagaG0youGLu6xjrtulGvuW+KZqgXHPvx9Hjj+tYnKc4Nss0ZNuBAfgwDGEgn1EbtrIsSWx22cgmCOABWJA0AkKQQDoBhRAY0hFCl5QSVECQ-QIWAIGoIx1aoA2ZFESA0i7WAMAUSWGi4Jh2DYMZSCQarNDGIxogY+rKvdlhPYAASODBMAAFljwt45Ib2zdfcNRFIL8IAmPXpCNpg+DDTpfZ8f2QED4PQ8q8PI+jsBKsr8MeAT1OMBwPuO6rnGfYIPPZEL782Gt8QgA)

Reply:
Well spotted! In this case, the DOM mock is correct, and [the tests fail if we give them a hyperscript-generated `select` rather than a manually constructed one.](https://flems.io/#0=N4IgtglgJlA2CmIBcB2ADAOgGwEYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQIAL54a9RswBW5XgKEimhAK7UebCAIAEYWgGt4AZXgJtACgButWOvgBKPcAA61PXrsAnPbTY2H1I9AF49K1ofAHN1BkFSDARqaLYWMNDwtBcAfldxPSRXOwd4IpLHcQ8vH3g2dR9PMCs3EFILeG1WvH9A4LxqrwMWkD4ABx0BbuL7RyLWnFbxJwHPIebW8cnqaeAKsr1WgCYllcGvDdGJ3R3uGdL5kFpT1aHhzeupu73Zg9a0F7nd5Xba7faPU6DJweKo7ah8DCkMadEbtSxsDD7aYaLTbKxOdxrPgjdRjKABeAhADuEDSejS8D0pBk3l+eigfD08LYelqAFp2jyGb0ghAAEbqIR6CCEeksRlozo8li0EKEPg8dSkbGabQ3fGuIH8aikHnAaJ1HqCgrhADyYrkSowFrYtqp1AACj5xvAfGwAJ4AEUpPB8EAmfB8VgAEgAVACyABlzOiAKIIeIYsbeth8APInqtLEgaFrLzG02s0rmQK+sJ6BFI-1WHF6gS2X4EoFeQUYHj2WBWNIQUg9falt7LIH2x3aDBQeDEajwL0+v3NuNJlNK9PwTMYbN5vP+guHEDFnqEt5eF1FF2va+C8q-GtCHwPt7COK+2hihBFIJHA-PQpzhN5fH8etDBMbdrAnIYILFKDjDMDprAWVp4K8CCeGQmC0LYKwcCw1k-CgPDUPREYThLDwgTADBamoBcow5TV9zFPgoH9HpaBIgB6fj-HnPgGPVTVSHxOiywbDtqzqN8+wHABhPhNDYJwMHgABHdR7Ek7IgWJWgRIY8ctN0-SRkWWiwPWRjhBYqw2O-QQME47iejFUsjLkxxX19JTYFgVT1M0nS9NgAySOJMVTMxTsLMiySMNs+iHOY31nI1VyMQ8ni9B4HyZOJfYAp8IKQrUwRwssqKrEMkqrB4eLzIiqzUuKt4GKYpyXI4riCqgLqhlKl8FMC-tgtCmqkqs4jfKgVrEva+rjkwwZQNAjwER8TR8W4EB+DAMYSF9UQxV-CxDsVNsTVEABOJAcAkKQQDoBhRD7UhFCO5RhDYUQJAAXQIWAIGoIxyCQKh3pkURIDSMNYAwBRDoaLgmHYNgxlIJBBM0MYjGiPtRP4xGWGRgABI4MEwAAWcnaUpkhUd+-NZFIUNw0ByRpE+pg+CRTp0Z8TGQGx3H8f4wnidJsB+KF5EeCp+mMBwdXFeFlq0YIDnRC5sMJhB8QgA)

Reply:
I'll have to dig into `m.render` and the mocks to try and figure out why the setter is triggered extraneously 
(tomorrow)

Reply:
In the hyperscript-generated vnode, the empty attrs object (i.e. {}) is re-set to null by execSelector(). This is the difference between the hyperscript-generated vnodes and the manually constructed ones in this case.

In execSelector(), instead of setting the attrs to null, it may be better to set an empty object or undefined.

Reply:
@kfule brilliant! And fixing this would lead to further standardisation. Thanks so much!